core/src/impl/Kokkos_CheckedIntegerOps.hpp

```diff
@@ -29,6 +29,11 @@ std::enable_if_t<std::is_integral_v<T>, bool> constexpr multiply_overflow(
     T a, T b, T& res) {
   static_assert(std::is_unsigned_v<T>,
                 "Operation not implemented for signed integers.");
+
+#if defined(__has_builtin) && __has_builtin(__builtin_mul_overflow)
+  return __builtin_mul_overflow(a, b, &res);
+#endif
```

Comment:
I think we should ifdef out the rest via a #else clause

Reply:
We have to use `#if` twice (to avoid errors on compilers that don't have `__has_builtin`), which would make the `#else` approach somewhat ugly.

This is also consistent with existing usage:
https://github.com/kokkos/kokkos/blob/1a3ea28f6e97b4c9dd2c8ceed53ad58ed5f94dfe/core/src/Kokkos_BitManipulation.hpp#L268-L274

Reply:
Are you saying this doesn't work: 

```c++
#if defined(__has_builtin) && __has_builtin(__builtin_mul_overflow)
  return __builtin_mul_overflow(a, b, &res);
#else
 ...
 #endif
 ```

Reply:
Yeah ok it doesn't work on ICC

Reply:
and we don't get unreachable code warnings?